# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [1]:
import math
def prvocisla(n):
    def test_prvocisla(j):
        test = True
        i = 2
        while i <= math.sqrt(j):
            if j % i == 0:
                test =False
                break
            else: 
                i += 1
        return test
    seznam_prvocisel = []
    k = 2
    while  len(seznam_prvocisel) < n:
        if test_prvocisla(k):
            seznam_prvocisel.append(k)
        k += 1
    return seznam_prvocisel


In [2]:
prvocisla(7)

[2, 3, 5, 7, 11, 13, 17]


# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [3]:
%%writefile modul_ze_tri_funkci.py
# Výpočet druhé odmocniny pomocí prostých iterací
def druha_odmocnina(cislo, pocet_iteraci):
    hodnota = cislo #odmocnovane cislo jako nulty clen iteracni posloupnosti
    for j in range(pocet_iteraci): #cyklus zajistujici vypocet odmocniny jako pevneho bodu kontraktivniho zobrazeni
        hodnota = (cislo/hodnota+hodnota)/2
    return hodnota
# Výpočet obsahu vepsaných úhelníků
def obsaheuler(pocet_iteraci):
    strana = 1  
    vyska =  druha_odmocnina(3, 100)/2 #vyska v sestiuhelniku
    for j in range(pocet_iteraci):
        strana = druha_odmocnina((strana/2)**2+(1-vyska)**2, 100) #vypocet zakladny
        vyska =  druha_odmocnina(1-(strana/2)**2, 100)
    return 6*(2**pocet_iteraci)*strana*vyska/2 #funkce vraci obsah vepsaneho 6*2**pocet_iteraci uhelnika
# Součet řady vedoucí k aproximaci pí
def obsahnewton(pocet_clenu):
    a_i=1/16 #stanoveni prvniho clenu
    soucet=(1/3)*a_i #soucet pro i=1
    for i in range(2,pocet_clenu+1):  
        a_i=(1/4)*a_i*(2*i-3)/(2*i) #vypocet nasledujiciho koeficientu
        soucet += (1/(2*i+1))*a_i   #dalsi scitanec k souctu rady
    return 12*(1/2-druha_odmocnina(3,1000)/8-soucet) #funkce vraci aproximaci cisla pi pomoci n clenu rady

Writing modul_ze_tri_funkci.py


In [4]:
import modul_ze_tri_funkci as mtrif
%time mtrif.druha_odmocnina(2,8)
%timeit mtrif.druha_odmocnina(2,8)

CPU times: total: 0 ns
Wall time: 0 ns
5.45 µs ± 1.48 µs per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [5]:
%%time
mtrif.obsaheuler(50)
mtrif.obsahnewton(50)

CPU times: total: 0 ns
Wall time: 3 ms


3.1415926535897936

In [6]:
%%timeit
mtrif.obsaheuler(50)
mtrif.obsahnewton(50)

2.23 ms ± 688 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)



# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [7]:
#m=1
#for j in round(1,m+1):
print(math.pi) 
print(round(abs(mtrif.obsaheuler(5)-math.pi),10))   
n=5
presnost=3.0
print(f"n={n},pi={mtrif.obsaheuler(n):.{round(abs(mtrif.obsaheuler(n)-math.pi))}f},pocet správných cifer={round(abs(mtrif.obsaheuler(n)-math.pi))}")
n=6
presnost=5
print(f"n={n},pi={mtrif.obsahnewton(n):.{round(abs(mtrif.obsaheuler(n)-math.pi))}f},pocet správných cifer={round(abs(mtrif.obsahnewton(n)-math.pi))}")

3.141592653589793
0.0005607027
n=5,pi=3,pocet správných cifer=0
n=6,pi=3,pocet správných cifer=0



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [ ]:
# kostra k doplnění, místa označená TODO doplňte

class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass
    
    def __add__(self, other):
        # TODO (použijte Addition)
        pass
    
    def __sub__(self, other):
        # TODO (použijte Addition a Multiplication se zápornou konstantu)
        pass
    
    def __mul__(self, other):
        # TODO (použijte Multiplication)
        pass
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
        self.argument = argument
    
    def __str__(self):
        return type(self).__name__ + "(" + str(self.argument) + ")"
        
class Addition(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
        pass
    def diff(self):
        # TODO
        pass
    
class Multiplication(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)*(pravy_argument)
        pass
    def diff(self):
        # TODO
        pass
    
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        # TODO vypíšeme něco jako value
        pass
    def diff(self):
        # TODO
        pass
    
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        # TODO vypíšeme něco jako x   
        pass
    def diff(self):
        # TODO
        pass
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        # TODO vypíšeme něco jako argument^exponent
        pass
    def diff(self):
        # TODO
        pass
    
class sin(UnaryOperator):
    def diff(self):
        # TODO
        pass

class cos(UnaryOperator):
    def diff(self):
        # TODO
        pass

class exp(UnaryOperator):
    def diff(self):
        # TODO
        pass
    
class ln(UnaryOperator):
    def diff(self):
        # TODO
        pass

class frc(UnaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako 1/(argument)
        pass
    
    def diff(self):
        # TODO
        pass

In [ ]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [ ]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [ ]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...